chainermn/communicators/pure_nccl_communicator.py

```diff
@@ -62,76 +65,113 @@ def _init_comms(self):
     def broadcast_data(self, model):
         _communication_utility.broadcast_naive(self.mpi_comm, model)
 
-    def allreduce_grad(self, model, stream=None):
-        self._init_comms()
-        if stream is None:
-            stream = chainer.cuda.Stream.null
+    def allreduce_grad(self, model):
+        stream = chainer.cuda.Stream.null
+        self.allreduce_grad_async(model, stream)
 
+    def allreduce_grad_async(self, model, stream):
+        self._init_comms()
         params = _memory_utility.extract_params(model)
         grad_dtype = _get_param_grad_dtype(params[0])
         if self.allreduce_grad_dtype is None:
             allreduce_grad_dtype = grad_dtype
         else:
             allreduce_grad_dtype = self.allreduce_grad_dtype
         n_elems = sum(param.grad.size for param in params)
-        self._assign(grad_dtype, allreduce_grad_dtype, n_elems)
-        self._pack_params_to_buffer(params, grad_dtype, allreduce_grad_dtype,
-                                    n_elems)
-        if stream != chainer.cuda.Stream.null:
+        needs_sync = self._assign(grad_dtype, allreduce_grad_dtype, n_elems)
+        if stream != chainer.cuda.Stream.null and needs_sync:
             chainer.cuda.Stream.null.synchronize()
```

Comment:
Why synchronization is required when double buffering is not enabled? I thought everything happens in null stream in that case... I want to know why and would be nice if the reason is documented here in the comment.

Reply:
This  synchronize() runs when double buffering is enabled and CUDA memory is allocated. So it is not required when double buffering is not enabled.